django/contrib/postgres/search.py

```diff
@@ -219,6 +221,53 @@ def as_sql(self, compiler, connection, function=None, template=None):
         return sql, extra_params + params
 
 
+class SearchHeadline(Func):
+    function = 'ts_headline'
+    output_field = TextField()
+    options = None
+
+    def __init__(self, expression, query, **extra):
+        if not hasattr(query, 'resolve_expression'):
+            query = SearchQuery(query)
+        self.config = query.config
+        self.options = extra.get('options', self.options)
+        super().__init__(expression, query, **extra)
+
+    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
+        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        if self.config:
+            if not hasattr(self.config, 'resolve_expression'):
+                resolved.config = Value(self.config).resolve_expression(query, allow_joins, reuse, summarize, for_save)
+            else:
+                resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
+        return resolved
```

Comment:
Good idea!

Reply:
I wonder if we could add a `SearchConfig` expression :thinking: (maybe a subclass of `Value`):
```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        ...

    def as_sql(self, compiler, connection):
        ...
        return '%::regconfig', params
```
This should simplify everything.

Reply:
I guess something like this could work:

```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        self.config = config

    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        if self.config:
            if not hasattr(self.config, 'resolve_expression'):
                resolved.config = Value(self.config).resolve_expression(query, allow_joins, reuse, summarize, for_save)
            else:
                resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved

    def as_sql(self, compiler, connection):
        sql, params = compiler.compile(self.config)
        sql = '{}::regconfig'.format(sql)
        return sql, params
```

Not really sure what the point of basing this class of `Value` would be.

Note that every class using `SearchConfig` would need to implement `resolve_expression` like this (please correct me if I'm missing something here):

```python
class SearchHeadline(Func):
    def __init__(self, expression, query, config=None, **extra):
        ...
        self.config = SearchConfig(config) if config else None

    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        if self.config:
            resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved
    
    ...
```

It could then be used in `as_sql` as something like this:
```python
def as_sql(self, compiler, connection, ...):
    ...
    if self.config:
        config_sql, config_params = compiler.compile(self.config)
        extra_context['config'] = '{}, '.format(config_sql)
```

As it still would require overriding `resolve_expression` (we still could use a mixin but simpler) it would still be a bit cumbersome but I guess that not having to handle the casting to `::regconfig` everywhere weight heavier. What do you think?

Reply:
I pushed a version which introduces `SearchConfig` as an expression (https://github.com/django/django/pull/12297/commits/709e36b5df769a169df7953cf0b54c81e4743b08). I prefer it over the mixin approach. Let me know what you think. If you agree I'll squash it into https://github.com/django/django/pull/12297/commits/96b55ce9c9c72d4f5fcde4f9eb85425426201266 and if not we can just revert it.

Reply:
@hannseman Thanks :+1: 

>  I prefer it over the mixin approach.

Yes me too :+1: . We can move `Value()` wrapping to the `__init__()` and simplify it a bit, e.g.:
```python
class SearchConfig(Expression):
    def __init__(self, config):
        super().__init__(output_field=None)
        if not (config and hasattr(config, 'resolve_expression')):
            config = Value(config)
        self.config = config
                          
    def resolve_expression(self, query=None, allow_joins=True, reuse=None, summarize=False, for_save=False):
        resolved = super().resolve_expression(query, allow_joins, reuse, summarize, for_save)
        resolved.config = self.config.resolve_expression(query, allow_joins, reuse, summarize, for_save)
        return resolved
        
    def as_sql(self, compiler, connection):
        sql, params = compiler.compile(self.config)
        return '%s::regconfig' % sql, params
```

Please move introducing a `SearchConfig` expression to the separate commit, or even PR.

Reply:
Much cleaner with the `Value` in `__init__()` 👍 

I opened up a new PR for this: https://github.com/django/django/pull/12377

I'll rebase this PR once https://github.com/django/django/pull/12377 is merged.

Reply:
Merged :confetti_ball:

Reply:
Rebased :raised_hands: